File: madameautomata-bot-python-autoreg/bot/commands/task.py

import discord
from discord import app_commands, Interaction
from ..theme import gothic_embed
from ..util.perm import is_domme, is_staff, is_sub
from ..db import Session, User, Collaring, Cage, JailSession, Star, Task, Worship, Config, CollarType, StarStatus
from ..util.time import parse_duration, random_case
from ..util.ids import cuid
from datetime import datetime, timedelta
from sqlalchemy import select, func, update, and_

def setup(bot):
    @bot.tree.command(name="task", description="Assign a task to a sub (role-wide TBD).")
    @app_commands.describe(sub="Assignee (pick a sub)", title="Short title", details="Details / checklist", deadline="Relative time like 12h or 2d")
    async def task(i: Interaction, sub: discord.User | None, title: str, details: str | None = None, deadline: str | None = None):
        member = i.guild.get_member(i.user.id)
        if not await is_domme(member):
            return await i.response.send_message("Only Dommes can use this.", ephemeral=True)

        due_at = None
        if deadline:
            import re, time
            m = re.fullmatch(r'(?:(\d+)d)?(?:(\d+)h)?(?:(\d+)m)?', deadline.strip(), re.I)
            if m:
                d,h,mi = (int(x) if x else 0 for x in m.groups())
                secs = (((d*24 + h)*60) + mi)*60
                if secs>0:
                    from datetime import datetime, timedelta
                    due_at = datetime.utcnow() + timedelta(seconds=secs)


if sub is None:
    # Role-wide blast to @taskslut
    async with Session() as s:
        cfg = await s.get(Config, 1)
    if not cfg or not cfg.taskslut_role_id:
        return await i.response.send_message("No TaskSlut role configured. Use /setup_set_role key:TaskSlut first.", ephemeral=True)
    role = i.guild.get_role(int(cfg.taskslut_role_id))
    if not role:
        return await i.response.send_message("TaskSlut role ID is invalid in this server.", ephemeral=True)
    members = [m for m in role.members if not m.bot]
    if not members:
        return await i.response.send_message("No human members have the TaskSlut role.", ephemeral=True)

    created = []
    async with Session() as s:
        for m in members:
            t = Task(id=cuid(), owner_id=i.user.id, assignee_id=str(m.id), title=title, details=details, due_at=due_at)
            s.add(t)
            created.append(t)
        await s.commit()

    # DM each sub a task with an Accept button
    accepted_count = 0
    for t in created:
        try:
            user = await i.client.fetch_user(int(t.assignee_id))
            view = discord.ui.View()
            async def make_cb(tid):
                async def accept_cb(interaction: discord.Interaction):
                    if interaction.user.id != int(t.assignee_id):
                        return await interaction.response.send_message("Only the assignee can accept.", ephemeral=True)
                    async with Session() as s2:
                        tt = await s2.get(Task, tid)
                        tt.accepted = True
                        await s2.commit()
                    await interaction.response.edit_message(embed=gothic_embed("Task Accepted", f"Task **{title}** accepted."), view=None)
                return accept_cb
            btn = discord.ui.Button(label="Accept Task", style=discord.ButtonStyle.success)
            btn.callback = await make_cb(t.id)
            view.add_item(btn)
            body = f"You were assigned: **{title}**\n{details or ''}"
            if due_at:
                body += f"\nDue: <t:{int(due_at.timestamp())}:R>"
            await user.send(embed=gothic_embed("New Task", body), view=view)
        except Exception:
            pass

    # Announce in channel & ping role
    announce = f"Assigned **{len(created)}** tasks to {role.mention}. Each sub received a DM with an Accept button."
    await i.response.send_message(embed=gothic_embed("Task Blast", announce))
    return

        async with Session() as s:
            t = Task(id=cuid(), owner_id=i.user.id, assignee_id=sub.id, title=title, details=details, due_at=due_at)
            s.add(t)
            await s.commit()
            task_id = t.id

        view = discord.ui.View()
        async def accept_cb(interaction: discord.Interaction):
            if interaction.user.id != sub.id:
                return await interaction.response.send_message("Only the assignee can accept.", ephemeral=True)
            async with Session() as s:
                tt = await s.get(Task, task_id)
                tt.accepted = True
                await s.commit()
            await interaction.response.edit_message(embed=gothic_embed("Task Accepted", f"Task **{title}** accepted by <@{sub.id}>."), view=None)

        btn = discord.ui.Button(label="Accept Task", style=discord.ButtonStyle.success)
        btn.callback = accept_cb
        view.add_item(btn)

        body = f"**<@{sub.id}>** {title}\n{details or ''}"
        if due_at:
            body += f"\nDue: <t:{int(due_at.timestamp())}:R>"
        await i.response.send_message(embed=gothic_embed("Task Assigned", body), view=view)

File: madameautomata-js-bot/src/commands/task.js
import {
  SlashCommandBuilder,
  PermissionFlagsBits,
  ChannelType,
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle,
  StringSelectMenuBuilder,
  EmbedBuilder
} from "discord.js";
import { prisma } from "../lib/prisma.js";
import { Theme, gothicEmbed } from "../theme/theme.js";
import { parseDuration, randomCase } from "../util/duration.js";
import { cuid } from "../util/ids.js";
import { isDomme, isSub, isStaff, getConfig } from "../util/perm.js";

const cmd = {
  data: new SlashCommandBuilder()
    .setName("task")
    .setDescription("Assign a task to a sub or blast to @taskslut (leave sub empty).")
    // REQUIRED FIRST
    .addStringOption(o =>
      o.setName("title").setDescription("Title").setRequired(true)
    )
    // THEN OPTIONALS
    .addUserOption(o =>
      o.setName("sub").setDescription("Assignee (leave empty for role blast)")
    )
    .addStringOption(o =>
      o.setName("details").setDescription("Details / checklist")
    )
    .addStringOption(o =>
      o.setName("deadline").setDescription("Relative time like 12h or 2d")
    ),

  async run(i) {
    if (!await isDomme(i.member)) {
      return i.reply({ content: "Only Dommes can use this.", ephemeral: true });
    }

    const sub = i.options.getUser("sub");
    const title = i.options.getString("title", true);
    const details = i.options.getString("details") || null;
    const deadline = i.options.getString("deadline");

    let dueAt = null;
    if (deadline) {
      const m = /^(?:(\d+)d)?(?:(\d+)h)?(?:(\d+)m)?$/i.exec(deadline.trim());
      if (m) {
        const d = Number(m[1] || 0), h = Number(m[2] || 0), mi = Number(m[3] || 0);
        const secs = (((d * 24 + h) * 60) + mi) * 60;
        if (secs > 0) dueAt = new Date(Date.now() + secs * 1000);
      }
    }

    // Single sub path
    if (sub) {
      const t = await prisma.task.create({
        data: { id: cuid(), ownerId: i.user.id, assigneeId: sub.id, title, details, dueAt }
      });

      const accept = new ButtonBuilder()
        .setCustomId(`accept:${t.id}`)
        .setStyle(ButtonStyle.Success)
        .setLabel("Accept Task");
      const row = new ActionRowBuilder().addComponents(accept);

      const body = `${title}\n${details || ""}${dueAt ? `\nDue: <t:${Math.floor(dueAt.getTime() / 1000)}:R>` : ""}`;

      try {
        const dm = await sub.createDM();
        await dm.send({ embeds: [gothicEmbed("Task Assigned", `**<@${sub.id}>** ${body}`)], components: [row] });
        return i.reply({ embeds: [gothicEmbed("Task Assigned", `Sent to <@${sub.id}> via DM.`)], ephemeral: true });
      } catch {
        return i.reply({ embeds: [gothicEmbed("Task Assigned", `**<@${sub.id}>** ${body}`)], components: [row] });
      }
    }

    // Role blast
    const cfg = await getConfig();
    if (!cfg.taskSlutRoleId) {
      return i.reply({ content: "No TaskSlut role set. Use /setup.", ephemeral: true });
    }
    const role = i.guild.roles.cache.get(cfg.taskSlutRoleId);
    if (!role) {
      return i.reply({ content: "Configured TaskSlut role not found on this server.", ephemeral: true });
    }

    const members = role.members.filter(m => !m.user.bot);
    let sent = 0;

    for (const m of members.values()) {
      const t = await prisma.task.create({
        data: { id: cuid(), ownerId: i.user.id, assigneeId: m.id, title, details, dueAt }
      });

      const accept = new ButtonBuilder()
        .setCustomId(`accept:${t.id}`)
        .setStyle(ButtonStyle.Success)
        .setLabel("Accept Task");
      const row = new ActionRowBuilder().addComponents(accept);

      const body = `${title}\n${details || ""}${dueAt ? `\nDue: <t:${Math.floor(dueAt.getTime() / 1000)}:R>` : ""}`;

      try {
        const dm = await m.createDM();
        await dm.send({ embeds: [gothicEmbed("Task Assigned", `**<@${m.id}>** ${body}`)], components: [row] });
        sent++;
      } catch {
        // ignore DM failures; they can accept from channel summary if you post it
      }
    }

    return i.reply({
      embeds: [gothicEmbed("Task Blast", `Assigned **${members.size}** tasks to <@&${role.id}>. DMs sent to **${sent}**.`)]
    });
  }
};

export default cmd;
